Delaying sending abnormal memory event send

kneeraj opened the pull request:
## Changes proposed in this pull request

Delaying the sending of abnormal memory event for some time as an inexplicable
spike in tenured heap space is being seen which is failing operations owing to
LowMemoryException. The sampler thread delegates this to a separate thread when it comes to poll the heap for the first time.

@rk-geode you also please look at the changes.
## Patch testing

tpcc perf test which was failing with this exception passes now even if spike comes.
## Other PRs

(Does this change required changes in other projects- snappyData, spark, spark-jobserver, aqp? Add the links of PR of the other subprojects that are related to this change)

…

Commit:
Delaying the sending of abnormal memory event for some time as an iexplicable
spike in tenured heap space is being seen which is failing operations owing to
LowMemoryException.
The sampler thread delegates this to a separate thread when it comes to poll the
heap for the first time.

Commit:
Fixed an issue while stoppping the thread.

Commit:
Fixed issues in dunit tests because of the change in behavior of delaying
memory events.
Fixed test to allow creation of index on lob columns LangScripts_LOBDUnit.java,
unrelated to this change.

Commit:
some more fixes to tests after recent change of allowing clob types to be
pk and/or index columns.

Commit:
Increased retry time span to fix intermittent failure in
testUseCase8HangStandaloneServer_5965_2

sumwale commented on gemfirexd/tools/src/dunit/java/com/pivotal/gemfirexd/query/JSONBasicTestsDUnit.java:
```diff
@@ -542,7 +542,8 @@ public void testAlterTable() throws Exception {
-  public void testUnsupportedClauses() throws Exception {
+  // This test should be removed as these restrictions have been removed
+  public void _testUnsupportedClauses() throws Exception {
     startVMs(1, 3, 0, null, null);
```
The valid tests in this is JSON as PK or index column. Partitioning is allowed, so those can be removed but other two can be retained.

kneeraj replied:
JSON has been explicitly restricted so commented out the test initially. Now activated it and modified the test as per expected exception.

sumwale commented on gemfirexd/tools/src/test/java/com/pivotal/gemfirexd/jdbc/LangScripts_LOBTest.java:
```diff
@@ -597,10 +597,10 @@ public void testLangScript_LOB_WithPartitioning() throws Exception {
-      { "create index id on d(a)", null },
-      { "create index ie on e(a)", null },
+      { "create index id on d(a)", "42832" },
+      { "create index ie on e(a)", "42832" },
```
LOB as PK or index key got explicitly disabled with the unified partitioner merge which had my patch for the same. Allowing this had problems because we have no handling of *CompositeKeys for byte[][]'s especially if the key is a combination of LOB and non-LOB (then it needs to pick up a combo of LOB byte[] and first byte[]).

Commit:
Adding multiple fixes for regression

Commit:
Picking row formatter based on byte value only. No optimization of latest schema now.

Commit:
Merge branch 'regression-tmp-branch' into SNAP-728

Commit:
Avoiding the extra thread for delaying memory event as per review comment.

Commit:
Reverting files whose changes are coming through other PRs.

Commit:
Corrected the delay flag.

Commit:
Made a couple of fixes to delaying of memory event.

Commit:
added temporary debug lines

Commit:
Delaying memory event base on time instead of counter as updateStateAndSendEvent method
is not only called by State sampler thread but Notification handler thread as well.

Commit:
Better fix for SNAP-728. Avoid LowMemoryException due to unexplained memory spike

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/catalog/ExtraTableInfo.java:
```diff
@@ -190,50 +190,56 @@ public final RowFormatter getRowFormatter(
+//    else {
+//      return this.container.getRowFormatter(vbytes);
+//    }
+    return this.container.getRowFormatter(vbytes);
```
Is it being tracked in JIRA as to why it was required to remove these blocks?

kneeraj replied:
The bug related to this has not been closed but deferred from 0.5

sumwale commented on gemfirexd/tools/src/dunit/java/com/pivotal/gemfirexd/ddl/GfxdLRUDUnit.java:
```diff
@@ -329,6 +331,10 @@ public String description() {
+  public static void setDummytestBytes() {
+    HeapMemoryMonitor.setTestBytesUsedForThresholdSet(100);
+  }
+
```
This is a static so needs to be cleared out in tearDown else will interfere with other tests (disable with -1).

kneeraj replied:
Good point. Thanks,

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/control/HeapMemoryMonitor.java:
```diff
@@ -612,6 +631,46 @@ public void updateStateAndSendEvent(long bytesUsed) {
+      boolean isDownBy10percentOrMore = false;
+      if ( prevBytesUsed != 0 && delta < 0) {
+        long percent = (-delta/prevBytesUsed) * 100;
+        isDownBy10percentOrMore = percent > 10 ? true : false;
```
unnecessary boolean if-else. Just use: percent > 10

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/control/HeapMemoryMonitor.java:
```diff
@@ -573,37 +573,56 @@ public final void updateStateAndSendEvent() {
+
+  private long prevBytesUsed = 0;
+  private int countSinceUpsurge = 0;
+  private long upsurgeStartTime = 0;
```
Can these have sync issues, or we know that all methods here are invoked in the same single thread?

kneeraj replied:
There is a big sync in the method "updateStateAndSendEvent" which calls this so yes one thread at a time. It is thread safe.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/control/HeapMemoryMonitor.java:
```diff
@@ -612,6 +631,46 @@ public void updateStateAndSendEvent(long bytesUsed) {
+        long percent = (-delta/prevBytesUsed) * 100;
+        isDownBy10percentOrMore = percent > 10 ? true : false;
+      }
+      if (isDownBy10percentOrMore) {
```
I think 10% decrease in just 1sec is unlikely. Perhaps it will be better to count prevBytesUsed since the last time countSinceUpsurge became > 0

sumwale replied:
Ok, this is checking if it is a genuine spike or a fake one, and for fake one a >10% increase, right? Looks fine then.

kneeraj replied:
Almost. It is looking for a change in 10% when the spike is coming down.

sumwale commented:
Overall looks good. See some comments.

Commit:
Some test cleanup changes as per review comment.
